File: backend/src/ssm/auth/htpasswd.py
"""Parse and verify Apache-style ``.htpasswd`` files.

Only bcrypt hashes are accepted (``$2a$``, ``$2b$``, ``$2y$``); other historical
Apache hash formats are rejected so a weak hash in the file cannot authenticate
anyone. Callers should run ``htpasswd -cB`` which emits the bcrypt variant.
"""

from __future__ import annotations

import logging
from pathlib import Path

import bcrypt

_log = logging.getLogger(__name__)
_BCRYPT_PREFIXES: tuple[str, ...] = ("$2a$", "$2b$", "$2y$")


def verify_password(password: str, hashed: str) -> bool:
    """Return True iff ``password`` matches the bcrypt ``hashed`` value.

    Non-bcrypt or malformed hashes yield False rather than an exception so a
    single bad line in ``.htpasswd`` can never grant access.
    """
    if not password or not hashed:
        return False
    if not hashed.startswith(_BCRYPT_PREFIXES):
        return False
    # bcrypt.checkpw requires $2a/$2b; normalize the Apache $2y$ variant.
    normalized = hashed.replace("$2y$", "$2b$", 1) if hashed.startswith("$2y$") else hashed
    try:
        return bcrypt.checkpw(password.encode("utf-8"), normalized.encode("utf-8"))
    except (ValueError, TypeError):
        return False


class HtpasswdStore:
    """In-memory cache of ``username → bcrypt hash`` loaded from disk."""

    def __init__(self, path: Path) -> None:
        self._path = path
        self._entries: dict[str, str] = {}
        self.reload()

    def reload(self) -> None:
        """Re-read the file. Missing or unreadable files leave the store empty."""
        self._entries = _parse_htpasswd(self._path)

    def list_users(self) -> list[str]:
        return sorted(self._entries)

    def verify(self, username: str, password: str) -> bool:
        hashed = self._entries.get(username)
        if hashed is None:
            return False
        return verify_password(password, hashed)


def _parse_htpasswd(path: Path) -> dict[str, str]:
    if not path.exists():
        return {}
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError as exc:
        _log.warning("could not read htpasswd %s: %s", path, exc)
        return {}

    entries: dict[str, str] = {}
    for lineno, raw_line in enumerate(raw.splitlines(), start=1):
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        if ":" not in line:
            _log.warning("skipping malformed htpasswd line %d in %s", lineno, path)
            continue
        username, _, hashed = line.partition(":")
        username = username.strip()
        hashed = hashed.strip()
        if not username or not hashed:
            _log.warning("skipping malformed htpasswd line %d in %s", lineno, path)
            continue
        entries[username] = hashed
    return entries

File: phoenix/lib/ssm/auth/htpasswd.ex
defmodule Ssm.Auth.Htpasswd do
  @moduledoc """
  Parse and verify Apache-style `.htpasswd` files — 1:1 port of
  backend/src/ssm/auth/htpasswd.py.

  Only bcrypt hashes are accepted (`$2a$`, `$2b$`, `$2y$`); other historical
  Apache hash formats are rejected so a weak hash in the file can never
  authenticate anyone. Operators create entries with `htpasswd -cB`.

  The file is read per call: it is tiny, the OS caches it, and a password
  change or new user takes effect on the next request without any reload
  machinery.
  """

  require Logger

  @bcrypt_prefixes ["$2a$", "$2b$", "$2y$"]

  @doc "True iff `password` matches the bcrypt `hashed` value."
  @spec verify_password(String.t(), String.t()) :: boolean()
  def verify_password(password, hashed)

  def verify_password(password, hashed)
      when password in [nil, ""] or hashed in [nil, ""],
      do: false

  def verify_password(password, hashed) do
    if String.starts_with?(hashed, @bcrypt_prefixes) do
      # Bcrypt.verify_pass requires $2a/$2b; normalize the Apache $2y$ variant.
      normalized =
        case hashed do
          "$2y$" <> rest -> "$2b$" <> rest
          other -> other
        end

      Bcrypt.verify_pass(password, normalized)
    else
      false
    end
  rescue
    # A malformed hash line must yield false, never an exception.
    _ -> false
  end

  @doc """
  Verify `username`/`password` against the htpasswd file at `path`.
  Runs a dummy bcrypt verification for unknown users so response timing does
  not reveal which usernames exist.
  """
  @spec verify(Path.t(), String.t(), String.t()) :: boolean()
  def verify(path, username, password) do
    case Map.fetch(parse(path), username) do
      {:ok, hashed} ->
        verify_password(password, hashed)

      :error ->
        Bcrypt.no_user_verify()
        false
    end
  end

  @doc """
  A short fingerprint of the user's current hash line, stored in the session
  at login: when the entry changes or disappears, every session of that user
  dies on its next request.
  """
  @spec entry_fingerprint(Path.t(), String.t()) :: {:ok, String.t()} | :error
  def entry_fingerprint(path, username) do
    case Map.fetch(parse(path), username) do
      {:ok, hashed} ->
        fp =
          :crypto.hash(:sha256, hashed)
          |> Base.encode16(case: :lower)
          |> binary_part(0, 16)

        {:ok, fp}

      :error ->
        :error
    end
  end

  @doc "Sorted list of usernames in the file."
  @spec list_users(Path.t()) :: [String.t()]
  def list_users(path), do: path |> parse() |> Map.keys() |> Enum.sort()

  @doc """
  Read the file into `%{username => hash}`. Missing or unreadable files yield
  an empty map; malformed lines are skipped with a warning (python parity).
  """
  @spec parse(Path.t()) :: %{String.t() => String.t()}
  def parse(path) do
    case File.read(path) do
      {:ok, raw} ->
        raw
        |> String.split(["\n", "\r\n"])
        |> Enum.with_index(1)
        |> Enum.reduce(%{}, fn {raw_line, lineno}, acc ->
          parse_line(String.trim(raw_line), lineno, path, acc)
        end)

      {:error, :enoent} ->
        %{}

      {:error, reason} ->
        Logger.warning("could not read htpasswd #{path}: #{inspect(reason)}")
        %{}
    end
  end

  defp parse_line("", _lineno, _path, acc), do: acc
  defp parse_line("#" <> _, _lineno, _path, acc), do: acc

  defp parse_line(line, lineno, path, acc) do
    case String.split(line, ":", parts: 2) do
      [username, hashed] when username != "" and hashed != "" ->
        Map.put(acc, String.trim(username), String.trim(hashed))

      _ ->
        Logger.warning("skipping malformed htpasswd line #{lineno} in #{path}")
        acc
    end
  end
end
